packages/ember-application/lib/system/application-instance.js

```diff
@@ -241,20 +241,22 @@ const ApplicationInstance = EngineInstance.extend({
 
     let router = get(this, 'router');
 
-    let handleResolve = () => {
-      // Resolve only after rendering is complete
+    let handleTransitionResolve = () => {
       return new RSVP.Promise((resolve) => {
-        // TODO: why is this necessary? Shouldn't 'actions' queue be enough?
-        // Also, aren't proimses supposed to be async anyway?
-        run.next(null, resolve, this);
+        // Resolve once rendering is completed. `router.handleURL` returns the transition (as a thennable)
+        // which resolves once the transition is completed, but the transition completion only queues up
+        // a scheduled revalidation (into the `render` queue) in the Renderer.
+        //
+        // This uses `run.schedule('afterRender', ....)` to resolve after that rendering has completed.
+        run.schedule('afterRender', null, resolve, this);
```

Comment:
Interesting!  I hadn't thought of that when reviewing this, but you are correct when `shouldRender` is false we can just `return this;` right away I believe.